site/content/posts/2024-06-05-githubidentityprovider.md

```diff
@@ -0,0 +1,191 @@
+for details about how to create a GitHub App or GitHub OAuth app.
+
+```yaml
+apiVersion: idp.supervisor.pinniped.dev/v1alpha1
+kind: GitHubIdentityProvider
+metadata:
+  name: my-github-provider
+  namespace: pinniped-supervisor
+spec:
+  allowAuthentication:
+    organizations:
+      policy: AllGitHubUsers
+  client:
+    secretName: my-github-provider-client-secret
+---
+apiVersion: v1
+kind: Secret
+type: "secrets.pinniped.dev/github-client"
+metadata:
+  name: my-github-provider-client-secret
+  namespace: pinniped-supervisor
+stringData:
+  clientID: <client-id-from-github>
+  clientSecret: <client-secret-from-github>
+```
+
+This GitHubIdentityProvider uses github.com (the default) and allows any user of github.com to authenticate.
+
+But wait a minute! Any user of github.com? Aren't there millions of users? Yes, there are.
+This simplest configuration example is great for a demo or for a Kubernetes cluster running on your laptop,
+but you may not want to use this for your enterprise's fleets of Kubernetes clusters.
+
+However, note that you could use the above GitHubIdentityProvider along with a policy on the FederationDomain
+to reject authentication for any user unless they belong to certain GitHub teams. For example:
+
+```yaml
+apiVersion: config.supervisor.pinniped.dev/v1alpha1
+kind: FederationDomain
+metadata:
+  name: my-federation-domain
+  namespace: pinniped-supervisor
+spec:
+  issuer: https://pinniped.example.com/my-issuer-path
+  identityProviders:
+    - displayName: "My GitHub IDP 🚀"
+      objectRef:
+        apiGroup: idp.supervisor.pinniped.dev
+        kind: GitHubIdentityProvider
+        name: my-github-provider
+      transforms:
+        expressions:
+          - type: policy/v1
+            expression: 'groups.exists(g, g in ["my-github-org/team1", "my-github-org/team2"])'
+            message: "Only users in certain GitHub teams are allowed to authenticate"
+```
+
+Now users must belong to one of the two teams configured above to be able to successfully authenticate.
+
+## Restricting authentication by GitHub organization membership
+
+Would you rather only allow members of certain GitHub organizations to authenticate? No problem, we've got you covered.
+Just make a small change to your GitHubIdentityProvider.
+
+```yaml
+apiVersion: idp.supervisor.pinniped.dev/v1alpha1
+kind: GitHubIdentityProvider
+metadata:
+  name: my-github-provider
+  namespace: pinniped-supervisor
+spec:
+  allowAuthentication:
+    organizations:
+      policy: OnlyUsersFromAllowedOrganizations
+      allowed:
+        - my-enterprise-organization # this is case-insensitive
+  client:
+    secretName: my-github-provider-client-secret
+```
+
+Now users must belong to the organization configured above to successfully authenticate.
+
+When multiple orgs are `allowed` then the user must belong to any one of those orgs.
+
+Want to further restrict auth by GitHub team membership? No problem, you can still use the technique shown above.
```

Comment:
Clarify what `technique shown above` means?

Reply:
I know it means the `policy/v1` expression above but we should clarify.

Reply:
Updated and force pushed.